pennylane/pauli/pauli_word_sparse_matrix.md

```diff
@@ -0,0 +1,119 @@
+
+> In scipy, $\texttt{val}$ is $\texttt{data}$, $\texttt{col}$ is $\texttt{indices}$, and $\texttt{row}$ is $\texttt{indptr}$.
+
+If we look at the CSR representation of Pauli matrices, we already see some clear patterns in their sparse strucutre:
+$$\begin{align*}
+I &= \begin{cases}\texttt{val}=\left[1, 1\right]\\ \texttt{col}=\left[0, 1\right]\end{cases} \\
+X &= \begin{cases}\texttt{val}=\left[1, 1\right]\\ \texttt{col}=\left[1, 0\right]\end{cases} \\
+Y &= \begin{cases}\texttt{val}=\left[-i, i\right]\\ \texttt{col}=\left[1, 0\right]\end{cases} \\
+Z &= \begin{cases}\texttt{val}=\left[1, -1\right]\\ \texttt{col}=\left[0, 1\right]\end{cases}
+\end{align*}$$
+$Z$ and $I$ have the same structure, and the same happens with $X$ and $Y$. All of them have the same $\texttt{row}=[0, 1, 2]$, as all the matrices have a unique NNZ element in every row.
+
+Let's look at a simple example to illustrate the properties of the tensor product of Pauli matrices. Consider the Pauli word $XZI$, which can be created in PennyLane with `qml.pauli.PauliWord({0: "X", 1: "Z", 2: "I"})`. The corresponding operator is obtained by taking the tensor product of its elements:
+$$
+XZI = \begin{pmatrix}0 & 1\\ 1 & 0\end{pmatrix}\otimes \begin{pmatrix}1 & 0\\ 0 & -1\end{pmatrix}\otimes \begin{pmatrix}1 & 0\\ 0 & 1\end{pmatrix}
+$$
+
+Starting from the right, our initial matrix is $I$ with $\texttt{val}=\left[1, 1\right]$ and $\texttt{col}=\left[0, 1\right]$ (let us forget about $\texttt{row}$ for now). If we apply $Z$, we obtain:
+$$ZI = \begin{pmatrix}1 & 0\\ 0 & -1\end{pmatrix}\otimes \begin{pmatrix}1 & 0\\ 0 & 1\end{pmatrix} = 
+\left(\begin{array}{cc|cc}
+1 & 0 & & \\ 
+0 & 1 & & \\ 
+\hline 
+& & -1 & 0 \\
+& & 0 & -1
+\end{array}\right).$$
+The resulting CSR representation is $\texttt{val}=\left[1, 1, -1, -1\right]$ and $\texttt{col}=\left[0, 1, 2, 3\right]$.
+
+```python
+import numpy as np
+import scipy
+
+I = np.array([[1, 0], [0, 1]])
+Z = np.array([[1, 0], [0, -1]])
+X = np.array([[0, 1], [1, 0]])
+
+ZI = np.kron(Z, I)
+ZI_CSR = scipy.sparse.csr_matrix(ZI)
+print(ZI)
+print(f"val: {ZI_CSR[ZI_CSR.nonzero()].tolist()[0]}\ncol: {ZI_CSR.indices}")
+
+[[ 1  0  0  0]
+ [ 0  1  0  0]
+ [ 0  0 -1  0]
+ [ 0  0  0 -1]]
+val: [1, 1, -1, -1]
+col: [0 1 2 3]
+```
+
+Every time we perform a tensor product with a new Pauli matrix, we double the matrix dimensions ($\mathbb{C}^{m\times m}\to \mathbb{C}^{2m\times2m}$). Given their nature, we always replicate the structure of our existing matrix in blocks either along the diagonal (like in this case) or the anti-diagonal (like in the next operation). This means we always have as many NNZ elements as columns, and there is always one per column and row (hence why we ignore $\texttt{row}$). Thus, our $\texttt{val}$ and $\texttt{col}$ arrays double in size with every operation and the effect of each matrix in our sparse representation is well defined and consistent.
```

Comment:
We could mention that `row` can be somewhat demanding to generate, and hence we cache it.

Reply:
Maybe put that after the sentence below (just saw it).

Reply:
Done! Thanks for the feedback 😄